_posts/2019-11-19-demo-fraud-detection.md

```diff
@@ -0,0 +1,220 @@
+
+Although most examples in the [documentation](https://ci.apache.org/projects/flink/flink-docs-stable/dev/api_concepts.html#define-keys-using-field-expressions) use specific fixed events' fields, nothing prevents us from extracting them in a more dynamic fashion, based on the specifications of the rules. For this we will need one additional preparation step before invoking the `keyBy` function.
+
+Let's look at a high level how our main processing pipeline might look like:
+
+```java
+DataStream<Alert> alerts =
+    transactions
+        .process(new DynamicKeyFunction())
+        .keyBy(/* some key selector */);
+        .process(/* actual calculations and alerting */)
+```
+
+Given a set of predefined rules in the first step of the processing pipeline, we would like to iterate over them and prepare every event to be dispatched to a respective aggregating instance. This is what is done in `DynamicKeyFunction`:
+
+```java
+public class DynamicKeyFunction
+    extends ProcessFunction<Transaction, Keyed<Transaction, String, Integer>> {
+   ...
+  /* Simplified */
+  List<Rule> rules = /* Rules that are initialized somehow.
+                        Details will be discussed in a separate section. */;
+
+  @Override
+  public void processElement(
+      Transaction event,
+      Context ctx,
+      Collector<Keyed<Transaction, String, Integer>> out) {
+
+      for (Rule rule :rules) {
+       out.collect(
+           new Keyed<>(
+               event,
+               KeysExtractor.getKey(rule.getGroupingKeyNames(), event),
+               rule.getRuleId()));
+      }
+  }
+  ...
+}
+``` 
+ `KeysExtractor.getKey()` uses reflection to extract required values of `groupingKeyNames` fields from events and combines them as a single concatenated String key, e.g `{beneficiaryId=25;payeeId=12}`.
+
+Notice that a wrapper class `Keyed` with the following signature was introduced as the output type of `DynamicKeyFunction`:  
+
+```java   
+public class Keyed<IN, KEY, ID> {
+  private IN wrapped;
+  private KEY key;
+  private ID id;
+
+  ...
+  public KEY getKey(){
+      return key;
+  }
+}
+```
+
+Where `wrapped` is the original Transaction event data, `key` is the result of using `KeysExtractor` and `id` is the ID of the Rule which caused the dispatch of the event according to this Rule's grouping logic.
+
+Events of this type will be the input to the `keyBy` function of the main processing pipeline. This allows us to use a simple lambda-expression in place of a [`KeySelector`](https://ci.apache.org/projects/flink/flink-docs-stable/dev/api_concepts.html#define-keys-using-key-selector-functions) as the final step of implementing dynamic data shuffle.
```

Comment:
It could, I just found it a bit easier to understand what is going on. Not sure about the article - shuffle sounds like an uncountable to me, similar to 'modulation' here https://iopscience.iop.org/article/10.1088/0031-9155/40/5/005